Joomla! Update: reinstall Joomla! and Update & Install features | seccond try

zero-24 opened the pull request:
Pull Request for Issue #9469.
#### Summary of Changes

Added features to Joomla! Update:
- **Reinstall.** If you're already on the latest version you are given the option to re-download the full update package and install it on top of your site. Useful if someone has made core hacks and you want to undo them.
- **Upload & Update.** Now that we have removed the ability to install Joomla! as an extension to itself the users with no FTP access and/or no outbound server Internet access (e.g. behind a corporate DMZ firewall) are unable to update. We now have the option to upload the update package. Since this is dangerous and could potentially be abused you need to re-enter your login information after the upload and before the uploaded package is installed. Failure to comply removes the uploaded file.
#### Testing Instructions
###### First test
- install 3.4.8 from https://github.com/joomla/joomla-cms/releases/tag/3.4.8
- go to your database and run this command (it changes the update server of the update component to my own hosted version)

```
UPDATE  `#__update_sites` SET  `location` =  'http://www.jah-tz.de/downloads/extensions/dev/com_joomlaupdate.xml' WHERE  `location` = 'http://update.joomla.org/core/extensions/com_joomlaupdate.xml';
```
- go to Extensions -> Update
- install the com_joomlaupdate Update
- Go to com_joomlaupdate
- notice the layout of the new Update
- switch to the Tab "Upload & Install"
- download: https://github.com/joomla/joomla-cms/releases/download/3.5.0/Joomla_3.5.0-Stable-Update_Package.zip
- Upload the 3.5.0 update package.
- watch the progress bar
- see that you updated to 3.5.0 stable now.
###### Note

Please notice that as the 3.5.0 stable package don't include the changes here the new overview is gone for now. But if you follow the seccond test it come back :smile: 
###### seccond test
- install 3.5.0 stable or staging (you can also use the result of the first test)
- install com_patchtester
- apply this fix of com_patchtester: joomla-extensions/patchtester#133
- apply this patch
- go to com_joomlaupdate
- notice the new layout
- try the reinstall feature
- try the upload and install freature
#### For the code reviewer

The changes between the code by @nikosdion and me can be found here: https://github.com/nikosdion/joomla-cms/compare/feature/joomlaupdate-reinstall...zero-24:feature/joomlaupdate-reinstall
#### Some other things

The needed update xml for com_joomlaupdate can be found here: https://github.com/joomla/update.joomla.org/pull/6

The com_joomlaupdate package can be found here (it is the same as on my server for updating):
http://www.jah-tz.de/downloads/download.php?file=com_joomlaupdate.zip&amp;folder=extensions/dev/
#### Limitations of using the com_joomlaupdate standalone. (edit)

You can't use the reinstall feature as the standalone package don't update the updater lib file please see: https://github.com/joomla/joomla-cms/pull/9612#issuecomment-202142160
#### Thanks

Thanks goes to @nikosdion @wilsonge and @mbabker and com_patchtester (as there we can find how we should create a update packages)

Commit:
Joomla! Update: Reinstall current version

Add type hints to Joomla! Update's controller

Commit:
Joomla! Update: Reinstall current version

Refactored JUpdater, made it possible to record the current version of an extension into the #__updates table

Commit:
Joomla! Update: Reinstall current version

Model and view template changes

Commit:
Merge remote-tracking branch 'nikosdion/staging' into feature/joomlaupdate-reinstall

Commit:
Merge remote-tracking branch 'joomla/staging' into feature/joomlaupdate-reinstall

Commit:
Oops, typo

Commit:
Bump Joomla! Update version

Commit:
Move alternative panels into separate view template layouts

Keeping it it simple and clean

Commit:
Reinstall is only supported in newer versions of Joomla!

When JUpdater::findUpdates has four (or more) parameters we can use the reinstall support by passing true in the fourth parameter (includeCurrent). Otherwise we have to call findUpdates with three parameters which only gives us available updates. This is necessary since Joomla! Update can run under older Joomla! 3.x releases, most notably 3.4, which lack the reinstall support.

Commit:
Upload form in a tab

Commit:
Typo

Commit:
Upload to temporary file

Commit:
Captive login validation

Commit:
Final step to install from file

Commit:
Fix FTP mode; add Hybrid mode

While the FTP credentials were applied, the file write method was ignored. That explains why a buch of users reported that Joomla! Update didn't work for them.

Since the included restore.php supports the Hybrid file write mode (direct file writes where possible, FTP only when the file is unwriteable) I added that option as well.

Commit:
Bug fixes

* Missing token
* Typo in JoomlaupdateControllerUpdate::captive()
* We must store the base name, not the full path to the captive file

wilsonge committed:
Code style fixes

wilsonge committed:
Add back weblinks file

Commit:
Merge pull request #1 from wilsonge/joomlaupdate

Minor Fixes

Commit:
Removing the bang from "Joomla!" in the language files per @brianteeman's advice

Commit:
Misplaced token check

Commit:
Not enough coffee

Commit:
Profoundly crap code to make our asinine PHP Code Sniffer rules happy

zero-24 committed:
bump version number and some CS

zero-24 committed:
Update default.php

zero-24 committed:
version numbers

zero-24 committed:
some CS

zero-24 committed:
CS

zero-24 committed:
CS

zero-24 committed:
set login center and some cs

zero-24 committed:
Update view.html.php

zero-24 committed:
Update updater.php

zero-24 committed:
Update default_upload.php

zero-24 committed:
Update en-GB.com_joomlaupdate.ini

zero-24 committed:
Update default_upload.php

zero-24 committed:
fix postgresql and use "IS NULL" insted of "= ``"

Following this: http://www.postgresql.org/message-id/3FE325BD.1040507@3times25.net we can use IS NULL for chekcing for null values in integer fields:

>Wei Weng wrote:
> Geoffrey wrote:
> 
>> How does one check for an unset value in an integer field?
>>
>> I've tried such things as:
>>
>> select ..... where intnumber = ''
>> select ...... where intnumber =  ?
>> select .....  where intnumber = NULL
>>
>> Thanks.
>>
> It is actually WHERE intnumber IS NULL. You don't use operator = to 
> compare with NULLs, you use IS.
>
>Thank you muchly....

zero-24 committed:
Update update.php

zero-24 committed:
use the api

zero-24 committed:
Current plan is to ship it with 3.5.2

zero-24 committed:
Update update.php

zero-24 committed:
Update default.php

zero-24 committed:
Update view.html.php

zero-24 committed:
Update view.html.php

zero-24 committed:
Update updater.php

zero-24 committed:
Create default.min.js

zero-24 committed:
Create encryption.min.js

zero-24 committed:
Create json2.min.js

zero-24 committed:
Create update.min.js

zero-24 committed:
use &amp; insted of &

zero-24 committed:
super admins can't reset or remind the password

zero-24 committed:
Update captive.php

zero-24 committed:
Update default_reinstall.php

zero-24 committed:
Update default_update.php

zero-24 committed:
Update default_reinstall.php

zero-24 committed:
Update default_update.php

zero-24 committed:
Create default_noupdate.php

zero-24 committed:
Update default.php

zero-24 committed:
Update en-GB.com_joomlaupdate.ini

zero-24 committed:
Update default_noupdate.php

zero-24 committed:
Update en-GB.com_joomlaupdate.ini

zero-24 committed:
Update default.php

zero-24 committed:
Update default_reinstall.php

zero-24 committed:
Update default_update.php

zero-24 committed:
Update en-GB.com_joomlaupdate.ini

zero-24 committed:
Update en-GB.com_joomlaupdate.ini

zero-24 committed:
Update default_noupdate.php

zero-24 committed:
Rename default_noupdate.php to default_nodownload.php

zero-24 committed:
Update default.php

zero-24 committed:
Update default_nodownload.php

zero-24 committed:
Update en-GB.com_joomlaupdate.ini

zero-24 committed:
Update default_nodownload.php

zero-24 committed:
Update default_upload.php

zero-24 committed:
Update view.html.php

zero-24 committed:
Update joomlaupdate.xml

zero-24 committed:
Update joomlaupdate.xml

zero-24 committed:
Update joomlaupdate.xml

zero-24 committed:
Update joomlaupdate.xml

mbabker commented on administrator/components/com_joomlaupdate/joomlaupdate.xml:
```diff
@@ -7,10 +7,15 @@
-	<version>3.0.0</version>
+	<version>3.5.2</version>
 	<description>COM_JOOMLAUPDATE_XML_DESCRIPTION</description>
+	<media destination="com_joomlaupdate" folder="media">
```
The media directory needs to be added to the [Ant build script](https://github.com/joomla/joomla-cms/blob/staging/build.xml#L142) for this component.

zero-24 replied:
I hope it is correct see: https://github.com/zero-24/joomla-cms/commit/df19cc3049ec5fd2399619134da5e490a21cf969

mbabker replied:
It works.  Just tried it.

zero-24 committed:
Add the media folder to the build.xml thanks @mbabker

zero-24 committed:
Update default_nodownload.php

zero-24 committed:
Update default_nodownload.php

zero-24 committed:
Update default_nodownload.php

alikon commented:
following second test at step : apply this pr
 got from com_patchtester
Error
The file marked for modification does not exist: build.xml

zero-24 commented:
@aliko that is because i have add the media folder to this pr but it don't exist in stable. Can you test my zip file? I'm going to report it to com_patchtester :)

alikon commented:
after install your com_joomlaupdate.zip  -> in Components ->Joomla! Update
Parse error: syntax error, unexpected '<' in C:\xampp\htdocs\pr\administrator\components\com_joomlaupdate\views\default\tmpl\default_nodownload.php on line 20

zero-24 committed:
Update default.php

zero-24 commented:
Yes @alikon i have fixed the repo and my local instance but not the container :(

Now i have done it see the zip above.

brianteeman commented:
First test
Everything worked as described but I never saw a files extracted (this might be unrelated to this PR)

![screen shot 2016-03-27 at 05 40 48](https://issues.joomla.org/uploads/1/b820dfa12c0854f8c30536bd9d813702.png)
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9612">issues.joomla.org/joomla-cms/9612</a>.</sub>

brianteeman commented:
Test two - same issue as @alikon with the build.xml so manually installed the component from https://github.com/joomla/joomla-cms/files/190845/com_joomlaupdate.zip

I dont see any "reinstall" option

I do see a Warning section which isnt very helpful<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9612">issues.joomla.org/joomla-cms/9612</a>.</sub>

brianteeman commented on administrator/language/en-GB/en-GB.com_joomlaupdate.ini:
```diff
@@ -55,4 +63,7 @@ COM_JOOMLAUPDATE_VIEW_UPDATE_DOWNLOADFAILED="Download of update package failed."
 COM_JOOMLAUPDATE_VIEW_UPDATE_FILESEXTRACTED="Files extracted"
 COM_JOOMLAUPDATE_VIEW_UPDATE_INPROGRESS="Updating your Joomla files. Please wait ..."
 COM_JOOMLAUPDATE_VIEW_UPDATE_PERCENT="Percent complete"
+COM_JOOMLAUPDATE_VIEW_UPLOAD_CAPTIVE_INTRO_BODY="Make sure that the update file you have uploaded comes from the official Joomla download page. Afterwards, please confirm that you want to install it by re-entering the login information for your site &bdquo;%s&ldquo; below."
```
Please use &quot;%s&quot; as that is correct for en-GB

brianteeman replied:
I did a PR against your fork for this

brianteeman committed:
update quote

this is correct for en-GB

richard67 commented:
I just tested and **have same results as Brian for test 2**, i.e. **not see the reinstall option**, also not after toggling the update channel or using the clear cache button.

Joomla! caching is off btw.

In addition I tested **FTP installation method** for the Upload & Install, and this **is broken** (as it was with Nik's PR, too).

I entered correct server and port and credentials in the ftp fields, same info as I have saved in global config.

What is strange: It is also broken for the Live Update, not only for Upload & Install, but it was definitely working in past, and since the Live Update is the old functionality I would not expect any change there.

Could my problems with FTP mode be related to the fact that I password-protected my test domain?

But this was not a problem before in 3.4.8 or 3.5.0. or ever in past, so I do not really think so.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9612">issues.joomla.org/joomla-cms/9612</a>.</sub>

zero-24 commented:
@brianteeman @alikon for com_patchtester you need to apply: https://github.com/joomla-extensions/patchtester/pull/133

I can have a look into the other things later today.

richard67 commented:
@zero-24 If I find time I will try to find out what goes wrong with the FTP method. If I find out something I'll let you know. I wish you a good Easter fest.

richard67 commented:
@zero-24 I added some error_log call before the calls of function "@ftp_connect" in restore.php and could see that "$this->host" and "$this->port" are empty strings when "@ftp_connect" is called in function "connect()" of class AKPostprocFTP.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9612">issues.joomla.org/joomla-cms/9612</a>.</sub>

richard67 commented:
@zero-24 When comparing code of original update component and yours, I cannot find an explanation for the problems with the ftp mode. Can it be that the ftp mode also failed all the time with the original update component, and it fell back to file copy mode so nobody ever noticed, and your PR just makes this visible now but not caused the error?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9612">issues.joomla.org/joomla-cms/9612</a>.</sub>

richard67 commented:
@zero-24 You can find my fix for making FTP mode work here: [https://github.com/zero-24/joomla-cms/pull/2](https://github.com/zero-24/joomla-cms/pull/2 "https://github.com/zero-24/joomla-cms/pull/2").

Can you check and if ok merge?

And if you understand why this 1 line which I removed made it not work?

I do not really understand that, I only saw it was not there in the original 3.5.0 update component.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9612">issues.joomla.org/joomla-cms/9612</a>.</sub>

richard67 commented:
@nikosdion Do you know why FTP mode works only after removing the one line as I proposed with [zero-24#2](https://github.com/zero-24/joomla-cms/pull/2 "https://github.com/zero-24/joomla-cms/pull/2") ?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9612">issues.joomla.org/joomla-cms/9612</a>.</sub>

nikosdion commented:
@richard67 It doesn't fix anything. It introduces a bug, causing Joomla! Update to always use direct file writes no matter what you select.

richard67 commented:
@nikosdion I see, you are right. Sorry for disturbing. Have to see if I can find out what goes wrong.

andrepereiradasilva commented on administrator/components/com_joomlaupdate/views/default/tmpl/default_upload.php:
```diff
@@ -0,0 +1,168 @@
+	</p>
+</div>
+
+<?php if (count($this->warnings != 0)) : ?>
```
something is not right here.

zero-24 replied:
Thnaks fixed with https://github.com/joomla/joomla-cms/pull/9612/commits/36b18498ea54d99f96b0cbf653d88e10a4f139fa

zero-24 committed:
Merge pull request #1 from brianteeman/patch-1

update quote

zero-24 committed:
Fixed count thanks @andrepereiradasilva

zero-24 commented:
>I do see a Warning section which isnt very helpful

Should be fixed with my last commit for the bug that @andrepereiradasilva found https://github.com/joomla/joomla-cms/pull/9612/commits/36b18498ea54d99f96b0cbf653d88e10a4f139fa

@alikon @brianteeman 

I have just checked the issue that you can't see the reinstall.

There we have two issues.
- With the package you installaed i can not update the lib file but that you need for reinstall
- And there is a problem with the "Test" Update Server (see: https://github.com/joomla/update.joomla.org/pull/7). If you use the standard Update server all works great.

##### Workarrounds for the reinstall

So we have two workarrounds.
First manually patch the lib file here (just replace with the new one): `libraries/joomla/updater/updater.php`
https://raw.githubusercontent.com/zero-24/joomla-cms/feature/joomlaupdate-reinstall/libraries/joomla/updater/updater.php

Seccond work arround is apply this patch using com_patchtester and this fix for com_patchtester: https://github.com/joomla-extensions/patchtester/pull/133

##### Limitations of the com_joomlaupdate standalone release

The only limitation should be that the reinstall not work as this requires the new lib file that can't (ok should not) be shipped with the standalone release.

zero-24 commented:
I have just updated the standalone package and fixed the testinstructions for reinstall if you come with the standalone package and not with the fixed com_patchtester

zero-24 commented:
@richard67 As for FTP i cannot help here. I have no setup that require ftp updates.

@alikon did you test with postgres? So we are sure that the SQL error is gone / fixed with this version of that branch as it should ;)

richard67 commented:
@zero-24 If you do not have a setup that **requires** ftp updates, you maybe have one which would **allow** them? It just needs an ftp server with old insecure passive ftp on your test system.

It seems FTP access information (hostname and so on) are not passed over, the variables are empty when calling ftp_connect.

It could be that this also does not work on staging or 3.5.0 or even 3.4.8 but nobody ever noticed because direct write is used silently in background. At leats it seemed like this when I tried my (wrong) PR for you before.

I have to check if adding some error_log call before the calls of function "@ftp_connect" in restore.php will show that "$this->host" and "$this->port" are empty strings when "@ftp_connect" is called in function "connect()" of class AKPostprocFTP also on a staging or a clean 3.5.0 (or even a 3.4.8), then we know for sure it is not related to this PR.

Beside this, everything worked in direct write mode here.

zero-24 committed:
add language file to the package in build.xml

zero-24 commented:
@mbabker can you check here: https://github.com/joomla/joomla-cms/pull/9612/commits/ea5f2496f72dccb5ccd24d951ac487362dbe2ca1 that i have not broken the ant script and the language files get copied to the right places?

brianteeman commented:
I still dont see any reinstall option. Maybe I am looking in the wrong place - can you post a screenshot of what I should see<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9612">issues.joomla.org/joomla-cms/9612</a>.</sub>

zero-24 commented:
@brianteeman 
- What workarround did you try? The new com_patchtester release or replaced the lib file manually?
- What Update server do you use?

The reinstall feature looks like this 
![reinstall_feature](https://cloud.githubusercontent.com/assets/2596554/14067922/620bf1c0-f472-11e5-8aed-fc7861d3819d.JPG)

brianteeman commented:
I had to press the clear cache button on that page
Before that it just didnt show

On 27 March 2016 at 22:20, zero-24 <notifications@github.com> wrote:

> @brianteeman <https://github.com/brianteeman>
>
>    - What workarround did you try? The new com_patchtester release or
>    replaced the lib file manually?
>    - What Update server do you use?
>
> The reinstall feature looks like this
> [image: reinstall_feature]
> <https://cloud.githubusercontent.com/assets/2596554/14067922/620bf1c0-f472-11e5-8aed-fc7861d3819d.JPG>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/9612#issuecomment-202147506>
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

Webdongle commented:
Unable to install patch tester in 3.4.8 to run the test

richard67 commented:
@Webdongle Testing instructions do not say anything about installing the patchtester on the 3.4.8. 
Beside this, the latest patch tester RC (and also Beta 5 before) require Joomla! 3.5.

zero-24 commented:
@Webdongle 

Patchtester requires 3.5 but you can test the first case without patchtester on the 3.5.0 you can install patchtester ;)

@brianteeman 
>I had to press the clear cache button on that page
>Before that it just didnt show

Ok so it works now for you?

Webdongle commented:
@zero-24 @richard67 

Test says install 3.4.8 ...

 I use patch tester to pull the pr

brianteeman commented:
yes all good except for my comment earlier about the files extracted


will try and test on pgsql before i go on holiday

On 27 March 2016 at 22:27, zero-24 <notifications@github.com> wrote:

> @Webdongle <https://github.com/Webdongle>
>
> Patchtester requires 3.5 but you can test the first case without
> patchtester on the 3.5.0 you can install patchtester ;)
>
> @brianteeman <https://github.com/brianteeman>
>
> I had to press the clear cache button on that page
> Before that it just didnt show
>
> Ok so it works now for you?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/9612#issuecomment-202147855>
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

zero-24 commented:
Great thanks @brianteeman !

@Webdongle 

>Test says install 3.4.8 ...

Yes but the first test says nothing about com_patchtester ;)

There you need to run the update SQL (to use my update server) and install the update for com_joomlaupdate than you can use the new layout and update to 3.5.0. :)

In the seccond test you start with a 3.5.0 and you can install com_patchtester :)

mbabker commented:
If you really need the patch tester, you can go back a couple of beta releases.  The beta 4 tag was the last one to support Joomla 3.3 and up.

moshiurshahin commented:
com_patchtester
Error this error facing.i am cpompletely unstable,i need to work dione.http://blogspotforbloggertipsandtricks.blogspot.com

Webdongle commented:
It's OK ... was too keen to test and never read the instructions carefully

Webdongle commented:
install 3.5.0 stable or staging (you can also use the result of the first test)
install com_patchtester
apply this fix of com_patchtester: joomla-extensions/patchtester#133 

#133 is there any need now it has been merged into patch tester ?

richard67 commented:
I have tested this item :white_check_mark: successfully on fd8e57664180761ace3e3da66ffa53554de8366d<br /><br />Everything works in direct write mode.

Problems with FTP mode are not related to this PR. I opened a separate issue for that.

What is not nice but I can live with it:

If you just have updated and so the installed version is the latest available, it needs again a click on the "Clear cache" button to make the "reinstall" option visible.

But I can live with that and so see my test as successful.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9612">issues.joomla.org/joomla-cms/9612</a>.</sub>

zero-24 commented:
@richard67 Thanks for testing :+1: 

>If you just have updated and so the installed version is the latest available, it needs again a click on the "Clear cache" button to make the "reinstall" option visible.

Just to explain the problem this only happens the first time as we change the updater.php file that adds some extra info to the update object. That extra info is not there bevor we update and the new code (and the new updater.php) runs only if we check again for updates or clear the update cache so this is expected :smile:

Webdongle commented:
I have tested this item :white_check_mark: successfully on fd8e57664180761ace3e3da66ffa53554de8366d<br /><br />OK ... read the instructions thoroughly now

Test 1 Full Success

Test 2 reinstall and Upload&Update Full Success

This is a fantastic well needed feature.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9612">issues.joomla.org/joomla-cms/9612</a>.</sub>

richard67 commented:
@zero-24 Well, yes, I simply did not have in mind that the updates to which I updated in the test did not contain your PR, so after update it had to be applied again and then needed also the clear cache button. With a Joomla! update container patched with this PR, the test would have been like in real life.

brianteeman commented:
tested with postgres - ok<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9612">issues.joomla.org/joomla-cms/9612</a>.</sub>

alikon commented:
@zero-24  with current staging second test with

-  this updater https://raw.githubusercontent.com/zero-24/joomla-cms/feature/joomlaupdate-reinstall/libraries/joomla/updater/updater.php

- Patch Tester 2.0.0 Release Candidate

still unable to see the reinstall button
![lateststaging](https://cloud.githubusercontent.com/assets/181681/14074552/0d806446-f4d2-11e5-97dd-f4259ce33ab9.PNG)
 what i'm doing wrong?

zero-24 commented:
You need to change the version.php so that you get JVERSION = 3.5.0 else the updater can't find a update or the files as 3.5.1-dev is not uploaded to the update server ;)

brianteeman commented:
That happened to me just now on a new test after it had worked before -odd

zero-24 commented:
@brianteeman which joomla version you use / have in the version.php the only version that can find the update package is 3.5.0

And if you use the Tst Update server there is a problem with that server (i have created a PR against that repo to fix it)

alikon commented:
chenged version.php file like 3.5.0 stable but still no reinstall button....
i'm on default channel

richard67 commented:
@alikon Because I see your screenshots shows you are on 3.5.1-dev: For testing with this 3.5.1-dev version you can use following **custom url**: [http://test5.richard-fath.de/list_test1.xml](http://test5.richard-fath.de/list_test1.xml), which I made for one of my PRs.

If offers also a 3.5.1-dev version, and with this I could see the "reinstall" thing because offered version = installer version.

@brianteeman If you also are aleady on 3.5.1.-dev, this is of course also possible for you.

No need to really install from my custom ULR then, but if you wanna helt with testing my PR (just look for open PRs from me, there are 2, the younger one is the one where this custom URL is used) would be nice.

richard67 commented:
@alikon Ah I see now you are on 3.5.0. Then my previous post will not help.

alikon commented:
@richard67 i've tried with a custom url but maybe mine was wrong... let's try yours i just want to see
the "Super-Usefull" reinstall button

andrepereiradasilva commented:
try this one as custom URL

https://raw.githubusercontent.com/andrepereiradasilva/update.joomla.org/master/www/core/teststs/list_sts.xml

And clear the cache

##### Additional info

In this case, the `list_sts.xml` xml is

```xml
<extensionset name="Joomla Core" description="Joomla! Core">
	<extension name="Joomla" element="joomla" type="file" version="3.5.1-dev" targetplatformversion="3.5" detailsurl="https://raw.githubusercontent.com/andrepereiradasilva/update.joomla.org/master/www/core/teststs/extension_sts.xml"/>
</extensionset>
```

The `extension_sts.xml` xml is

```xml
<?xml version="1.0" ?>
<updates>
	<update>
		<name>Joomla! 3.5.1</name>
		<description>Joomla! 3.5.1 CMS</description>
		<element>joomla</element>
		<type>file</type>
		<version>3.5.1-dev</version>
		<infourl title="Joomla!">https://www.joomla.org/announcements/release-news/5654-joomla-3-5-is-here.html</infourl>
		<downloads>
			<downloadurl type="full" format="zip">https://github.com/joomla/joomla-cms/releases/download/3.5.0/Joomla_3.5.0-Stable-Update_Package.zip</downloadurl>
		</downloads>
		<tags>
			<tag>stable</tag>
		</tags>
		<maintainer>Joomla! PLT</maintainer>
		<maintainerurl>https://www.joomla.org</maintainerurl>
		<section>STS</section>
		<targetplatform name="joomla" version="3.[345]" />
		<php_minimum>3.10.0</php_minimum>
	</update>
</updates>
```

richard67 commented:
Same as mine regarding XMLs (exccept the zip behind it).

andrepereiradasilva commented:
i think for 3.5.0 is just replace the `3.5.1-dev` parts by `3.5.0` in the two xmls.

richard67 commented:
yep. Often the 1st one is forgotten to change. "targetplatform" is also important, has to be as shown in your xml above, i.e. inlcude the 5 in the subversion pattern.

alikon commented:
:star2::star2: mine custom url .xml file was wrong with yours finally reinstall button appears :tada:

brianteeman commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9612">issues.joomla.org/joomla-cms/9612</a>.</sub>

richard67 commented:
@brianteeman Why can't it go into 3.5.1?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9612">issues.joomla.org/joomla-cms/9612</a>.</sub>

brianteeman commented:
That decision is nothing at all to do with me. It is up to the maintainer
(@wilsonge)

I am guessing because it has new language strings.

richard67 commented:
Ah, OK. Just wantd to ask, not wanted to blame anyone.

Makes sense with the language strings, but on the other hand we had so many issues with people trying to use extension installer hat I thought it would be better to have it in 3.5.1 even if not translated yet. But I might be wrong of course.

brianteeman commented:
I agree with you but it's not my call

zero-24 commented:
The last plan by @wilsonge i know is a standalone release of com_joomlaupdate this way this improvents (expected the reinstall feature) can be installed also in 3.4.8 for updating ;)

zero-24 commented:
For that reason i have moved the since tags to 3.5.2 too :)

infograf768 commented:
>The last plan by @wilsonge i know is a standalone release of com_joomlaupdate this way this improvents (expected the reinstall feature) can be installed also in 3.4.8 for updating ;)

But what would happen for new lang strings? TTs need these to be in core en-GB

richard67 commented:
If George does not use at least 1 of my changes in my PR for solving problems on updating related to the utf8mb4 conversion, he will have to release a 3.5.2 anyway soon after 3.5.1, but that's another story.

Webdongle commented:
The Update component was originally provided free by Nick from his Extension.  If the Update component is now to be a 'Stand alone' has Nick been consulted ?

wilsonge commented:
The update component was made standalone since Joomla 3.5.0 - me and nic have been in extensive discussions about how to roll this out since the decision was made  (and this pull request he made was made at my request) so yes he has been consulted

Actually since the mess of 3.3.4->3.3.6 Joomla Update has been able to be updated standalone - we have just not needed to do it yet. And yes it comes at the price of TT's not being able to translate the strings in < 3.5.2 so it's not a decision that I'm taking lightly either - but it's there for people who want it and those who have already upgraded to the latest stable version of Joomla have it available as well

infograf768 commented:
@wilsonge
If joomlaupdate is now standalone and new strings will not be in core en-GB until 3.5.2, we still can propose TTs who can do to include the new strings into their 3.5.1 packs to come. Those using crowdin will not be able for sure, but others can.

Are the strings proposed here in this PR sure to be added? If yes, just let me know and I will let it know to TTs.

Webdongle commented:
Would it not make more sense to have it as part of the Update component for 3.5.0 onwards and a separate standalone for older versions ?
If it's a stand alone only then it will (in essence to the ordinary user ) be no different than any other 3rd party extension.  But if it is part of the core then it will be a feature.  @wilsonge please reconsider.

>  me and nic have been in extensive discussions about how to roll this out since the decision was made
And he is no longer adding code to it.  I hope he doesn't see a stand alone version as 'competition' to his extensions. [at the risk of being banned for expressing an opinion]  imho it is Nick who should decide if it it is to be a stand alone instead of part of Joomla core.

brianteeman commented:
My understanding is that it IS part of the core but by technically making
it standalone AND offering it on its own it can
a) be installed on earlier releases
b) upgraded outside of the joomla core update process (this is really
important)

On 29 March 2016 at 12:05, Kevin Griffiths <notifications@github.com> wrote:

> Would it not make more sense to have it as part of the Update component
> for 3.5.0 onwards and a separate standalone for older versions ?
> If it's a stand alone only then it will (in essence to the ordinary user )
> be no different than any other 3rd party extension. But if it is part of
> the core then it will be a feature. @wilsonge
> <https://github.com/wilsonge> please reconsider.
>
> me and nic have been in extensive discussions about how to roll this out
> since the decision was made
> And he is no longer adding code to it. I hope he doesn't see a stand alone
> version as 'competition' to his extensions. [at the risk of being banned
> for expressing an opinion] imho it is Nick who should decide if it it is to
> be a stand alone instead of part of Joomla core.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/9612#issuecomment-202817725>
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

Webdongle commented:
Hi @brianteeman 

> My understanding is that it IS part of the core
I couldn't find it.  I clicked clear cache but the Joomla update component didn't offer to reinstall.  Nor was there a second tab to upload/install.  The update component in 3.5.1RC looks exactly like in 3.5.0 not like it did after I tested this patch.

wilsonge commented:
We are not going to be offering this on JED as "competition" to nic's stuff. It's just there so people who need file uploads and for some reason are still running on 3.4.8 can use file upload without having to run through the tedious file extraction method....

Webdongle commented:
@wilsonge 

Where is the stand alone update component ?  can I have a link please ?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9612">issues.joomla.org/joomla-cms/9612</a>.</sub>

wilsonge commented:
The "standalone" component is exactly what we ship in the CMS right now - you can generate the standalone package by running this script https://github.com/joomla/joomla-cms/blob/staging/build.xml#L142 - I haven't had time to do it myself yet so i can't link you to a package

Webdongle commented:
Sorry I don't understand.  How can it be the same ?  The update component in Joomla won't reinstall Joomla like the patch we tested.

I don't know how to run https://github.com/joomla/joomla-cms/blob/staging/build.xml#L142 to produce a zip to install into Joomla.  Sorry you don't have time to do it.  If I had not wasted so much time testing the patch by @zero-24 then I might have time to learn how to  run https://github.com/joomla/joomla-cms/blob/staging/build.xml#L142

Please, please, please, next time you decide not to merge a patch could you please tell us testers before we before we test it instead of us wasting our time testing something that won't be merged ?  Thanks

brianteeman commented:
Whoever said it wont be merged. This is RTC with a milestone of 3.5.2

wilsonge commented:
This as is is going to be merged. All we are going to do is also produce a zip of exactly this that people can upload onto older Joomla sites (especially people still stuck on 3.4.8)

Webdongle commented:
Thanks @brianteeman @wilsonge 

I obviously misunderstood.... I thought the stand alone zip was instead of having it in the core.

brianteeman commented:
No it will be in core but by making it something that can be installed it
can be installed in earlier releases of joomla but more importantly it
allows for the situation where we need to change the updater BEFORE someone
can update their joomla release.

On 7 April 2016 at 22:11, Kevin Griffiths <notifications@github.com> wrote:

> Thanks @brianteeman <https://github.com/brianteeman> @wilsonge
> <https://github.com/wilsonge>
>
> I obviously misunderstood.... I thought the stand alone zip was instead of
> having it in the core.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/9612#issuecomment-207090120>
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

mbabker commented:
For reference we made it possible to do this after 3.3.6 needed to be released because of an issue introduced into the restore.php script.  The mechanism has never been needed though.

rdeutz commented:
@zero-24 could you have a look at the merge conflicts, danke

zero-24 committed:
fix merge conflicts

zero-24 commented:
Fixed @rdeutz but please double check with @wilsonge about this PR.

richard67 commented:
I will be able to test again tomorrow if still necessary then.

andrepereiradasilva commented on administrator/language/en-GB/en-GB.com_joomlaupdate.ini:
```diff
@@ -29,11 +29,11 @@ COM_JOOMLAUPDATE_VIEW_COMPLETE_MESSAGE="Your site has been successfully updated.
-COM_JOOMLAUPDATE_VIEW_DEFAULT_FTP_PASSWORD="FTP password"
-COM_JOOMLAUPDATE_VIEW_DEFAULT_FTP_PORT="FTP port"
-COM_JOOMLAUPDATE_VIEW_DEFAULT_FTP_USERNAME="FTP username"
+CCOM_JOOMLAUPDATE_VIEW_DEFAULT_FTP_DIRECTORY="FTP Root"
```
CCOM?

brianteeman replied:
@zero-24 see https://github.com/zero-24/joomla-cms/pull/3

andrepereiradasilva commented on administrator/language/en-GB/en-GB.com_joomlaupdate.ini:
```diff
@@ -12,47 +12,58 @@ COM_JOOMLAUPDATE_CONFIG_UPDATESOURCE_CUSTOM_ERROR="The custom URL field is empty
-COM_JOOMLAUPDATE_VIEW_DEFAULT_FTP_DIRECTORY="FTP Root"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_NO_DOWNLOAD_URL="We can't find a download URL"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_NO_DOWNLOAD_URL_DESC="An update to Joomla %1$s was found, but it wasn't possible to fetch the download URL for that update. There are two possibilities why this happens:<br>- Your host doesn't support <a href="_QQ_"https://www.joomla.org/technical-requirements.html"_QQ_">the minimum requirements for Joomla %1$s</a> and there is no alternative download for your config ...
+CCOM_JOOMLAUPDATE_VIEW_DEFAULT_FTP_DIRECTORY="FTP Root"
```
CCOM?

brianteeman committed:
remove CCOM and replace with COM

zero-24 committed:
Merge pull request #3 from brianteeman/patch-2

remove CCOM and replace with COM

zero-24 commented:
Sorry guys i have just fixed it by merging brian's PR thanks @andrepereiradasilva & @brianteeman
